pkg/logger/logger.go

```diff
@@ -0,0 +1,328 @@
+	debug(1, l, msg, fields...)
+}
+
+// Info
+func info(skip int, l *Logger, msg string, fields ...Field) {
+	if !l.checkEnabledAndUpdateMetrics(skip + 1) {
+		return
+	}
+
+	l.l.Info(msg, fields...)
+}
+
+func Info(msg string, fields ...Field) {
+	info(1, defaultLogger, msg, fields...)
+}
+
+func (l *Logger) Info(msg string, fields ...Field) {
+	info(1, l, msg, fields...)
+}
+
+// Warn
+func warn(skip int, l *Logger, msg string, fields ...Field) {
+	if !l.checkEnabledAndUpdateMetrics(skip + 1) {
+		return
+	}
+
+	l.l.Warn(msg, fields...)
+}
+
+func Warn(msg string, fields ...Field) {
+	warn(1, defaultLogger, msg, fields...)
+}
+
+func (l *Logger) Warn(msg string, fields ...Field) {
+	warn(1, l, msg, fields...)
+}
+
+// Error
+func err(skip int, l *Logger, msg string, fields ...Field) {
+	if !l.checkEnabledAndUpdateMetrics(skip + 1) {
+		return
+	}
+
+	l.l.Error(msg, fields...)
+}
+
+func Error(msg string, fields ...Field) {
+	err(1, defaultLogger, msg, fields...)
+}
+
+func (l *Logger) Error(msg string, fields ...Field) {
+	err(1, l, msg, fields...)
+}
+
+// DPanic
+func dPanic(skip int, l *Logger, msg string, fields ...Field) {
+	if !l.checkEnabledAndUpdateMetrics(skip + 1) {
+		return
+	}
+
+	l.l.DPanic(msg, fields...)
+}
+
+func DPanic(msg string, fields ...Field) {
+	dPanic(1, defaultLogger, msg, fields...)
+}
+
+func (l *Logger) DPanic(msg string, fields ...Field) {
+	dPanic(1, l, msg, fields...)
+}
+
+// Panic
+func panicLogger(skip int, l *Logger, msg string, fields ...Field) {
+	if !l.checkEnabledAndUpdateMetrics(skip + 1) {
+		return
+	}
+
+	l.l.Panic(msg, fields...)
+}
+
+func Panic(msg string, fields ...Field) {
+	panicLogger(1, defaultLogger, msg, fields...)
+}
+
+func (l *Logger) Panic(msg string, fields ...Field) {
+	panicLogger(1, l, msg, fields...)
+}
+
+// Fatal
+func fatal(skip int, l *Logger, msg string, fields ...Field) {
+	if !l.checkEnabledAndUpdateMetrics(skip + 1) {
+		return
+	}
+
+	l.l.Fatal(msg, fields...)
+}
+
+func Fatal(msg string, fields ...Field) {
+	fatal(1, defaultLogger, msg, fields...)
+}
+
+func (l *Logger) Fatal(msg string, fields ...Field) {
+	fatal(1, l, msg, fields...)
+}
+
+// Sync
+func (l *Logger) Sync() error {
+	return l.l.Sync()
+}
+
+// Default returns default logger
+func Default() *Logger {
+	return defaultLogger
+}
+
+// package level Logger
+var defaultLogger = New(os.Stdout, InfoLevel)
+
+func New(w io.Writer, lvl Level, packages ...string) *Logger {
+	if w == nil {
+		panic("logger writer must be set")
+	}
+
+	var (
+		cfg     zap.Config
+		encoder zapcore.Encoder
+		metrics bool
+	)
+
+	if lvl == DebugLevel {
+		cfg = zap.NewDevelopmentConfig()
+		encoder = zapcore.NewConsoleEncoder(cfg.EncoderConfig)
+		metrics = true // enabled only for DebugLevel
+	} else {
+		cfg = zap.NewProductionConfig()
+		encoder = zapcore.NewJSONEncoder(cfg.EncoderConfig)
+	}
+
+	core := zapcore.NewCore(
+		encoder,
+		zapcore.AddSync(w),
+		zapcore.Level(lvl),
+	)
+
+	pkgs := map[string]bool{}
```

Comment:
Can you please explain how this is used? Maybe add usage to another package to show this mechanism off.

Reply:
I demonstrated in https://github.com/aquasecurity/tracee/pull/2093?ts=8#issuecomment-1227636610 comment. In tracee I think it could be useful when passing a flag like `tracee -debug=ebpf`.